crates/bevy_ecs/src/reflect/world.rs

```diff
@@ -0,0 +1,171 @@
+use bevy_reflect::{Reflect, ReflectFnsTypeData, ReflectFromPtr};
+
+use crate::component::ComponentId;
+use crate::entity::Entity;
+use crate::prelude::{AppTypeRegistry, World};
+
+impl World {
+    /// Retrieves an immutable `dyn T` reference to the given entity's Component of the given [`ComponentId`]
+    #[allow(dead_code)]
+    pub fn get_dyn_by_id<T: ReflectFnsTypeData>(
+        &self,
+        entity: Entity,
+        component_id: ComponentId,
+    ) -> Option<&T::Dyn> {
+        let Some(type_id) = self
+            .components()
+            .get_info(component_id)
+            .and_then(|n| n.type_id())
+        else {
+            return None;
+        };
+
+        let Some(dyn_obj) = self.get_dyn_reflect_by_id(entity, component_id) else {
+            return None;
+        };
+
+        let type_registry = self.resource::<AppTypeRegistry>();
+        let type_registry = type_registry.read();
+        type_registry
+            .get_type_data::<T>(type_id)
+            .and_then(|n| n.get(dyn_obj))
+    }
+
+    /// Retrieves an mutable `dyn T` reference to the given entity's Component of the given [`ComponentId`]
+    #[allow(dead_code)]
+    pub fn get_dyn_mut_by_id<T: ReflectFnsTypeData>(
+        &mut self,
+        entity: Entity,
+        component_id: ComponentId,
+    ) -> Option<&mut T::Dyn> {
+        let Some(type_id) = self
+            .components()
+            .get_info(component_id)
+            .and_then(|n| n.type_id())
+        else {
+            return None;
+        };
+
+        let type_registry = self.resource::<AppTypeRegistry>().clone();
```

Comment:
I don't understand why this needs to be cloned here.

Reply:
We should be able to just clone the much smaller `type_data`, which then drops the borrow on the `World` in time for it to be used in the `get_dyn_reflect_mut_by_id` call.